File: cpp/dsa/algorithms/arrays/product-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        // 1️⃣ Initialize result array
        vector<int> res(n, 1);
        
        // 2️⃣ First pass: Calculate prefix products
        int prefix = 1;
        for (int i = 0; i < n; i++) {
            res[i] = prefix;
            prefix *= nums[i];
        }
        
        // 3️⃣ Second pass: Calculate suffix products
        int suffix = 1;
        for (int i = n - 1; i >= 0; i--) {
            res[i] *= suffix;
            suffix *= nums[i];
        }
        return res;
    }
};

File: go/dsa/algorithms/arrays/product-except-self.go
func productExceptSelf(nums []int) []int {
    n := len(nums)
    // 1️⃣ Initialize result array
    res := make([]int, n)
    
    // 2️⃣ First pass: Calculate prefix products
    prefix := 1
    for i := 0; i < n; i++ {
        res[i] = prefix
        prefix *= nums[i]
    }
    
    // 3️⃣ Second pass: Calculate suffix products
    suffix := 1
    for i := n - 1; i >= 0; i-- {
        res[i] *= suffix
        suffix *= nums[i]
    }
    return res
}